// expresscheckout-nodejs/dist/cjs/Juspay.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Juspay = void 0;
const fs = __importStar(require("fs"));
const path_1 = __importDefault(require("path"));
const JuspayEnvironment_js_1 = require("./JuspayEnvironment.js");
const JuspayLogger_js_1 = require("./JuspayLogger.js");
const JuspayError = __importStar(require("./JuspayError.js"));
// resources
const Customer_js_1 = __importDefault(require("./resources/Customer.js"));
const Order_js_1 = __importDefault(require("./resources/Order.js"));
const Payments_js_1 = __importDefault(require("./resources/Payments.js"));
const OrderSession_js_1 = __importDefault(require("./resources/OrderSession.js"));
/**
 * Juspay's sdk
 */
class Juspay {
    constructor(juspayConfig) {
        const juspayEnvironment = new JuspayEnvironment_js_1.JuspayEnvironment(juspayConfig);
        this.juspayEnvironment = juspayEnvironment;
        this.order = new Order_js_1.default(juspayEnvironment);
        this.customer = new Customer_js_1.default(juspayEnvironment);
        this.payments = new Payments_js_1.default(juspayEnvironment);
        this.orderSession = new OrderSession_js_1.default(juspayEnvironment);
    }
    static getInstanceFromJuspayEnvironment(juspayEnvironment) {
        return new Juspay(juspayEnvironment.getJuspayConfig());
    }
    getJuspayEnvironment() {
        return this.juspayEnvironment;
    }
    static getSdkPackage() {
        if (Juspay.pkg != undefined) {
            return Juspay.pkg;
        }
        else {
            const filePath = path_1.default.join(process.cwd(), 'package.json');
            if (fs.existsSync(filePath)) {
                // eslint-disable-next-line @typescript-eslint/no-var-requires
                try {
                    const pkg = JSON.parse(fs.readFileSync(filePath).toString());
                    if (pkg.version == undefined || pkg.name == undefined) {
                        throw new JuspayError.JuspayError(`Critical! failed to read version or name in package.json, Please report this bug`);
                    }
                    else {
                        return { name: pkg.name, version: pkg.version };
                    }
                }
                catch (ignored) {
                    throw new JuspayError.JuspayError(`Critical! failed to parse package.json, Please report this bug.`);
                }
            }
            throw new JuspayError.JuspayError(`Critical! failed to read package.json, searched path:- ${filePath}, Please report this bug.`);
        }
    }
}
exports.Juspay = Juspay;
/** Http Request Defaults constants */
Juspay.DEFAULT_REQUEST_TIMEOUT = 80000;
Juspay.SANDBOX_BASE_URL = 'https://sandbox.juspay.in';
Juspay.PRODUCTION_BASE_URL = 'https://api.juspay.in';
Juspay.API_VERSION = '2019-05-07';
Juspay.pkg = Juspay.getSdkPackage();
Juspay.silentLogger = JuspayLogger_js_1.DefaultJuspayLogger.getSilentLogger();
/** exposing errors as static */
Juspay.errors = JuspayError;
exports.default = Juspay;
__exportStar(require("./JuspayLogger.js"), exports);
__exportStar(require("./JuspayEnvironment.js"), exports);
__exportStar(require("./JuspayError.js"), exports);
__exportStar(require("./types/index.js"), exports);
